### DISCUSSION TITLE
Implement `EVP_MD_FLAG_*FIPS*`?

### DISCUSSION THREAD

**Discussion by ngie-eign** [openssl/openssl#24643] [Q&A] [answered]
A few projects are currently relying on `EVP_MD_FLAG_*FIPS*` to be functional in order to allowlist certain operations, e.g., ntp tries to make use of the flag/feature when calculating a checksum as part of the protocol ([source](https://bk.ntp.org/ntp-dev/sntp/crypto.c?PAGE=anno&REV=660f2079Z83PNgpU9HLfQYT-US7vPg)) (search for `EVP_MD_CTX_FLAG_NON_FIPS_ALLOW`).

A similar change needs to be added to krb5 to handle [RFC-3961](https://www.rfc-editor.org/rfc/rfc3961):

> This encryption and checksum type pair is based on the Triple DES
> cryptosystem in Outer-CBC mode and on the HMAC-SHA1 message
> authentication algorithm.

Allowing the HMAC-SHA1 call would unbreak joining a Kerberized Active Directory domain on a FIPS 140-* enabled host using OpenSSL 3.x.

In short, having a documented interface which allows non-FIPS 140-* compatible EVP_MD_\* APIs to be used will enable a number of software libraries and applications to function without resorting to a third-party library like [libmd](https://man.freebsd.org/cgi/man.cgi?query=sha&sektion=3&apropos=0&manpath=FreeBSD+14.1-RELEASE) for computing checksums.

**Comment by paulidale** [comment, marked as answer]
Losing this flag was intentional, it's not coming back that I foresee.

The way to get non approved algorithms is via properties or by using a second library context.  E.g. load the FIPS and default providers and set the default property query to `?fips` will prefer FIPS implementations but not fail if it has to fall back to non approved ones.

BTW: HMAC-SHA1 is supported by the FIPS provider.  I suspect it's the 3DES part that's not.

**Reply by ngie-eign** [reply]
Ah, yes... Sorry -- 3DES.

**Reply by paulidale** [reply]
Try looking at the documentation for how to load multiples providers.  Eg. https://www.openssl.org/docs/manmaster/man7/fips_module.html

**Reply by ngie-eign** [reply]
Yes, that's fair.. I suppose I was hoping for an easy button alternative to solve this (I just stumbled upon the code in ntp and had an "aha!" moment about it being usable elsewhere), but I can see that creating and managing an additional `OSSL_LIB_CTX` is ultimately what needs to be done (it just seems really verbose).

**Reply by paulidale** [reply]
The rational behind this decision was to provide libcrypto with complete independence from all knowledge about FIPS.  The FIPS provider is meant to contain **all** of the FIPS knowledge.  OpenSSL 1.0.2 and the FOM didn't have this segregation which was a mistake.

Creating a new library context isn't that bad, it's a more lines of code than just setting a flag though.  Call `OSSL_LIB_CTX_new()` then add providers using either `OSSL_LIB_CTX_load_config()` or `OSSL_PROVIDER_load()`.

The project recommends that everything that uses OpenSSL should create its own private library context to operate within.  This prevents any possibility of cross contamination between usages (e.g. two different libraries or a library and the calling application).  The default library context is there for backwards compatibility and I expect it will be deprecated and removed eventually, but not in the near future.
